crates/tbdex/src/http/exchanges.rs

```diff
@@ -0,0 +1,319 @@
+    pub reply_to: Option<String>,
+}
+impl JsonDeserializer for CreateExchangeRequestBody {}
+impl JsonSerializer for CreateExchangeRequestBody {}
+
+#[derive(Serialize, Debug, PartialEq)]
+#[serde(untagged)]
+pub enum WalletUpdateMessage {
+    Order(Arc<Order>),
+    Cancel(Arc<Cancel>),
+}
+impl JsonSerializer for WalletUpdateMessage {}
+
+impl<'de> Deserialize<'de> for WalletUpdateMessage {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct MessageVisitor;
+
+        impl<'de> Visitor<'de> for MessageVisitor {
+            type Value = WalletUpdateMessage;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("an Order or Cancel")
+            }
+
+            fn visit_some<D>(self, deserializer: D) -> std::result::Result<Self::Value, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                let value: serde_json::Value = Deserialize::deserialize(deserializer)?;
+
+                let kind_str = value
+                    .get("metadata")
+                    .and_then(|m| m.get("kind"))
+                    .and_then(|k| k.as_str());
+
+                match kind_str {
+                    Some(k) => match MessageKind::from_str(k) {
+                        Ok(kind) => match kind {
+                            MessageKind::Order => {
+                                if let Ok(order) = serde_json::from_value::<Order>(value.clone()) {
+                                    Ok(WalletUpdateMessage::Order(Arc::new(order)))
+                                } else {
+                                    Err(serde::de::Error::custom("failed to deserialize order"))
+                                }
+                            }
+                            MessageKind::Cancel => {
+                                if let Ok(cancel) = serde_json::from_value::<Cancel>(value.clone())
+                                {
+                                    Ok(WalletUpdateMessage::Cancel(Arc::new(cancel)))
+                                } else {
+                                    Err(serde::de::Error::custom("failed to deserialize cancel"))
+                                }
+                            }
+                            _ => Err(serde::de::Error::custom(format!(
+                                "unexpected message kind {:?}",
+                                kind
+                            ))),
+                        },
+                        Err(_) => Err(serde::de::Error::custom(format!(
+                            "unexpected message kind {}",
+                            k
+                        ))),
+                    },
+                    None => Err(serde::de::Error::custom(format!(
+                        "unexpected message kind {:?}",
+                        kind_str
+                    ))),
```

Comment:
3 levels of unexpected message kind?

Reply:
hehe life ain't easy

Reply:
1. if the value is null
2. if the value cannot be parsed into the `MessageKind` enum (not a valid enum value)
3. if the kind is not one of the kind valid for the `WalletUpdateMessage` which is a subset of the overall `MessageKind`

life ain't easy lol